pxtblocks/fields/field_melodySandbox.ts

```diff
@@ -0,0 +1,407 @@
+            this.doneButton.addEventListener("click", () => this.onDone());
+            this.playButton = document.createElement("button");
+            pxt.BrowserUtils.addClass(this.playButton, "ui icon button");
+            this.playButton.id = "melody-play-button";
+            this.playButton.addEventListener("click", () => this.togglePlay());
+            this.playIcon = document.createElement("i");
+            this.playIcon.id = "melody-play-icon";
+            pxt.BrowserUtils.addClass(this.playIcon, "play icon");
+            this.playButton.appendChild(this.playIcon);
+            this.tempoDiv = document.createElement("div");
+            this.tempoDiv.id = "melody-tempo-div";
+            this.tempoLabel = document.createElement("label");
+            this.tempoLabel.id = "melody-tempo-label";
+            this.tempoLabel.innerText = "Tempo";
+            this.tempoInput = document.createElement("input");
+            pxt.BrowserUtils.addClass(this.tempoInput, "ui input");
+            this.tempoInput.type = "number";
+            this.tempoInput.value = this.tempo + ""; // will be updated according to slider
+            this.tempoInput.id = "melody-tempo-input";
+            this.tempoInput.addEventListener("input", () => this.setTempo(+this.tempoInput.value));
+            this.tempoDiv.appendChild(this.tempoLabel);
+            this.tempoDiv.appendChild(this.tempoInput);
+            this.bottomDiv.appendChild(this.tempoDiv);
+            this.bottomDiv.appendChild(this.playButton);
+            this.bottomDiv.appendChild(this.doneButton);
+            div.appendChild(this.bottomDiv);
+
+            // create event listeners at the end because the DOM needs to finish loading
+            for (let i = 0; i < this.numRow; i++) {
+                for (let j = 0; j < this.numCol; j++) {
+                    let el = "cell-" + i + "-" + j;
+                    document.getElementById(el).addEventListener("click", () => this.onNoteSelect(el));
+                }
+            }
+
+        }
+
+        // Runs when the editor is closed by clicking on the Blockly workspace
+        protected onEditorClose() {
+            if (this.isPlaying) {
+                this.stopMelody();
+            }
+        }
+
+        // when click done
+        onDone() {
+            if (this.isPlaying) {
+                this.stopMelody();
+            }
+            Blockly.DropDownDiv.hideIfOwner(this);
+        }
+
+        // This is the string that will be inserted into the user's TypeScript code
+        protected getTypeScriptValue(): string {
+            if (this.melody) {
+                return "\"" + this.melody.getStringRepresentation() + "\"";
+            }
+            return "";
+
+        }
+
+        // This should parse the string returned by getTypeScriptValue() and restore the state based on that
+        protected parseTypeScriptValue(value: string) {
+            value = value.slice(1, -1); // remove the boundary quotes
```

Comment:
user might decide to use ' or ` if editing in typescript.

Reply:
Do you have to unescape the script? May you simply need to do ``JSON.parse(value)``

Reply:
@pelikhan  What do you mean by refer?

Reply:
since value comes from typescript; it may have a different syntax so your parser needs to gracefully handle it. But we can harden that later. What happen when you have a variable name instead of a string literal?